lib/error_catalog.cpp

```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2018-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <map>
+#include <string>
+#include "error_catalog.h"
+#include "error_reporter.h"
+
+std::map<int, const std::string> ErrorReporter::errorCatalog = {
+    { ErrorType::ERR_UNKNOWN,     "%1%: Unknown %2%"},      // args: %1% - IR node, %2% - what
```

Comment:
I was hoping each catalog can be a complete sentence about the error and then developer can provide more reasoning on the specific occurrence of the error. The error message we print should be like a git commit message, with a title and a body. Title is mostly fixed template by the compiler, body is provided by the developer.

For example:

```
Error: "file.p4", line xxx, column xxx: (error code)
  Unsupported primitive clone_egress_to_egress
  The supported primitives are clone_ingress_to_egress, clone_ingress_to_ingress.
```

Reply:
Han's proposal would also make it easier to localize the errors if we decide to do it.

Reply:
I think this is reasonable. One potential issue is that we end up with too many errors codes, which will be confusing for developers to pick. So some amount of customization for the title will be needed as well. That said, I agree with giving it a try once we hear from a few more folks. 

I hope that it is clear what problem we're trying to address. Additional suggestions on how to make this work for both users and developers are more than welcome.